### ISSUE TITLE
static pods timeout on image pulling during cluster bringup

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#26626]
A few snaphosts from kubelet logs on problematic node from http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gke-large-cluster/31

```
I0601 11:14:10.933470    3467 kube_docker_client.go:252] Pulling image "gcr.io/google_containers/fluentd-gcp:1.18": "31b0cd3cfaa9: Downloading [>                                                  ] 524.6 kB/74.95 MB"
I0601 11:14:12.622702    3467 server.go:959] GET /healthz: (28.584µs) 200 [[curl/7.26.0] 127.0.0.1:54346]
...
I0601 11:21:10.933445    3467 kube_docker_client.go:252] Pulling image "gcr.io/google_containers/fluentd-gcp:1.18": "92ec6d044cb3: Downloading [====>                                              ] 6.301 MB/65.68 MB"
I0601 11:21:13.201412    3467 server.go:959] GET /healthz: (52.53µs) 200 [[curl/7.26.0] 127.0.0.1:54487]
```
(Roughly 1MB/min)

```
I0601 11:42:30.933474    3467 kube_docker_client.go:252] Pulling image "gcr.io/google_containers/fluentd-gcp:1.18": "31b0cd3cfaa9: Downloading [===============>                                   ] 22.57 MB/74.95 MB"
I0601 11:42:34.914121    3467 server.go:959] GET /healthz: (42.01µs) 200 [[curl/7.26.0] 127.0.0.1:54840]
I0601 11:42:39.519003    3467 server.go:959] GET /containerLogs/kube-system/fluentd-cloud-logging-gke-gke-large-cluster-default-pool-1-2ac9437d-caep/fluentd-cloud-logging: (289.353µs) 400
goroutine 9990 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc82107b340, 0x190)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa3
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc82107b340, 0x190)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x2b
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Response).WriteHeader(0xc82115b4a0, 0x190)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/response.go:200 +0x41
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Response).WriteErrorString(0xc82115b4a0, 0x190, 0xc8203681e0, 0x9a, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/response.go:180 +0x11a
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Response).WriteError(0xc82115b4a0, 0x190, 0x7f8e93825028, 0xc820da6e30, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/response.go:165 +0x8d
k8s.io/kubernetes/pkg/kubelet/server.(*Server).getContainerLogs(0xc8205c9040, 0xc82081a4b0, 0xc82115b4a0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/server/server.go:491 +0x14e9
k8s.io/kubernetes/pkg/kubelet/server.(*Server).(k8s.io/kubernetes/pkg/kubelet/server.getContainerLogs)-fm(0xc82081a4b0, 0xc82115b4a0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/server/server.go:344 +0x34
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc82074ae10, 0x7f8e93862f38, 0xc82107b340, 0xc820e3d340)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:272 +0xf30
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).(k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x7f8e93862f38, 0xc82107b3 [[Go-http-client/1.1] 10.240.1.110:36279]
...
I0601 11:53:20.933428    3467 kube_docker_client.go:252] Pulling image "gcr.io/google_containers/fluentd-gcp:1.18": "31b0cd3cfaa9: Downloading [=====================>                             ] 32.55 MB/74.95 MB"
I0601 11:53:22.509327    3467 server.go:959] GET /healthz: (26.907µs) 200 [[Go-http-client/1.1] 127.0.0.1:55018]
```

cc @wojtek-t @zmerlynn @kubernetes/goog-node @dchen1107

**Comment by zmerlynn** [comment]
cc @mattmoor

**Comment by gmarek** [comment]
It happened again in http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gke-large-cluster/33
again for fluentd...

**Comment by wojtek-t** [comment]
@yujuhong @Random-Liu - if you have any thoughts on it

**Comment by dchen1107** [comment]
Since we already have a static pod to re-pull test images, why not include fluentd one? @bprashanth

**Comment by wojtek-t** [comment]
I though fluentd is already there - isn't it?

**Comment by yujuhong** [comment]
> Since we already have a static pod to re-pull test images, why not include fluentd one? @bprashanth

We start all the add-on pods (e.g., fluentd, the image prepull pod) around the same time. Pulling fluentd in the prepull pod doesn't help.

**Comment by Random-Liu** [comment]
> I0601 11:14:10.933470    3467 kube_docker_client.go:252] Pulling image "gcr.io/google_containers/fluentd-gcp:1.18": "31b0cd3cfaa9: Downloading [>                                                  ] 524.6 kB/74.95 MB"
I0601 11:14:12.622702    3467 server.go:959] GET /healthz: (28.584µs) 200 [[curl/7.26.0] 127.0.0.1:54346]

The logs show that kube_docker_client is downloading different layers (the part before colon is layer id), so the throughput should > 1MB/s.

**Comment by girishkalele** [comment]
This run also had an error at the end about rate limiting preventing deletion of instances.
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gke-large-cluster/33

```
Deleting cluster gke-large-cluster...
                                     
Deleting cluster gke-large-cluster...done.
ERROR: (gcloud.container.clusters.delete) Some requests did not succeed:
 - Operation [<Operation
 name: u'operation-1464795621082-c963c468'
 operationType: OperationTypeValueValuesEnum(DELETE_CLUSTER, 2)
 selfLink: u'https://staging-container.sandbox.googleapis.com/v1/projects/51872839970/zones/us-central1-b/operations/operation-1464795621082-c963c468'
 status: StatusValueValuesEnum(DONE, 3)
 statusMessage: u'Google Compute Engine: Rate Limit Exceeded.'
 targetLink: u'https://staging-container.sandbox.googleapis.com/v1/projects/51872839970/zones/us-central1-b/clusters/gke-large-cluster'
 zone: u'us-central1-b'>] finished with error: Google Compute Engine: Rate Limit Exceeded.

2016/06/01 09:16:39 e2e.go:218: Error running teardown: exit status 1
2016/06/01 09:16:39 e2e.go:214: Step 'teardown' finished in 1h36m35.987410373s
```

**Comment by zmerlynn** [comment]
@girishkalele: That error is understood and being worked.

**Comment by bprashanth** [comment]
Any volunteers to trimming that image down?
```
10:06:43-beeps~/goproj/src/k8s.io/kubernetes] (lb_e2e)$ docker images | grep fluent
gcr.io/google_containers/fluentd-gcp                1.18                 ed70ff4ab587        12 weeks ago        411.5 MB
```

My first thought was to bake the install script into a busybox base and use sh instead of bash. Unfortunately only a handful of platforms are supported by the logging agent (https://cloud.google.com/logging/docs/agent/, why?), so the next best option is to use debian:jessie. 

There are probably a handful of small base images we can try. If all else fails we can build a custom rootfs like we do for the loadbalancer controllers (https://github.com/kubernetes/contrib/tree/master/images/ubuntu-slim).

**Comment by bprashanth** [comment]
@a-robinson as maintainer on the dockerfile

**Comment by yujuhong** [comment]
Trimming down the image will definitely help.

By the way, I missed the part that this happened on GKE, where the image prepull pod wasn't enabled. There was no pre-pulling at all. 

Another interesting thing is that the slow pull happened *after* all other nodes had already finished pulling their images, so it wasn't directly caused by other pulls in this cluster.

**Comment by girishkalele** [comment]
I see only about half of the node logs were uploaded to GCS.
build-log.txt mentions 2000 nodes, but I see only 1001 node folders under artifacts/.

**Comment by zmerlynn** [comment]
@yujuhong: What's stopping image prepull on GKE?

**Comment by bprashanth** [comment]
The static pod that prepulls e2e images isn't deployed on gke because it uses a salt boolean: https://github.com/kubernetes/kubernetes/blob/dc8ff81f5a48511f61ed2d65c0b8e745de03d09c/cluster/gce/config-test.sh#L165

**Comment by zmerlynn** [comment]
@spxtr for missing node logs in GCS

**Comment by zmerlynn** [comment]
Unfortunately, @roberthbailey and I have some very strong opinions on test configurations diverging this strongly from prod configurations, and this is huge. This is an option I actually considered probably 6mo-1y ago, but I rejected it because pre-pulling images is hugely cheating compared to how a customer would interact with the cluster.

A way to satisfy my objection is this: For every test configuration divergence, you need to create a *new* test configuration that includes the plain vanilla production configuration. You have to pay attention to both, though. As it is, on GCE we're now no longer testing image pulls under load (i.e. while doing other work) in any meaningful way - why is that not a useful test as well?

I can plumb this into GKE pretty easily, but we've pretty much rejected prod/test divergence on the GKE side, and I'm not sure we plan to start.

**Comment by wojtek-t** [comment]
In a separate thread I suggest pre-installing fluentd image in GCI. That would obviously not solve all problems of the world, but hopefully it would reduce flakiness. WDYT?

**Comment by a-robinson** [comment]
@bprashanth - I can give image trimming a try after 1.3, but probably not before. I'm having enough trouble figuring out how to get the proper dependencies for the journald plugin in place. Ruby dependency management is far from my expertise, unfortunately.

**Comment by zmerlynn** [comment]
@wojtek-t: It would help the GCI case, certainly.

**Comment by bprashanth** [comment]
> Unfortunately, @roberthbailey and I have some very strong opinions on test configurations diverging this strongly from prod configurations, and this is huge. This is an option I actually considered probably 6mo-1y ago, but I rejected it because pre-pulling images is hugely cheating compared to how a customer would interact with the cluster.

The pros and cons of image pulling didn't seem worth it:
1. registry and docker pull flake are nothing to do with kubernetes
2. we're still testing image pulls in node e2e, just in a more controlled environment
3. we don't know the conditions we want to test anyway. 

> As it is, on GCE we're now no longer testing image pulls under load (i.e. while doing other work) in any meaningful way - why is that not a useful test as well?

IMO we weren't doing this to begin with. Simply saying: it should work for some random images we throw at docker is insufficient. It makes us feel more secure, that's all. If we want to stress test image pulls, we should do that independently with a list of test cases that represent what we want to ensure. 

We have a curated list of images at hand. We don't have any idea if this is representative of what customers will be using in terms of base layers, image size etc. We don't know when and why it hangs. It is a net time sink for me (I'm guessing the team at large, but I did it because I have 1.3 priorities and this was distracting me from them) to keep debugging the same issues time and again. 

See: https://github.com/kubernetes/kubernetes/issues/25277, https://github.com/kubernetes/kubernetes/issues/26390 and email on the same topic where we discussed registry issues for more context.

**Comment by gmarek** [comment]
@bprashanth - I can't agree with you - registry and docker pull flakes have a lot to do with the kubernetes. We should have a mechanism that detects the problems and retries download - especially for system pods (I have a very strong opinion on divergence as well - exactly the same as I had with master registration)

@wojtek-t - I think that adding fluentd to the image may make it too heavy. Remember that this fluentd will be unnecessarily downloaded to every single machine.

**Comment by bprashanth** [comment]
we already retry, it times out by pushing the pod into exponential backoff. Sure we should implement workarounds that are smarter about why the pulling flaked etc, but we don't have those today.

**Comment by bprashanth** [comment]
we don't prepull system pods

**Comment by gmarek** [comment]
Do we retry after it times out, or do we have some mechanism to determine if it's already taking *too long* (for some definition of too long).

**Comment by vishh** [comment]
Testing the overall image pipeline is critical for production Kubernetes deployments, but is it critical for all Kubernetes PRs? 
As for kube PRs we primarily care about the image pulling logic on the kubelet side. 
By having dedicated tests for the kubelet logic and creating a separate test suite that pulls images from registries, can we switch the merge queue blocking test suites to use locally cached images?

**Comment by gmarek** [comment]
Yeah, but shouldn't the system handle this cases not only in test scenarios?

**Comment by vishh** [comment]
Kubernetes does handle image pipeline issues AFAIK. If we want some
reasonable timeouts for our merge queue blocking tests, we need to reduce
some of the dependencies.

**Comment by zmerlynn** [comment]
Don't enable it in `config-test.sh` blanketly if that's the case. If I had my druthers, there would be very little in `config-test.sh` that was different than `config-default.sh`.

**Comment by zmerlynn** [comment]
(It's easy to explicitly enable it on merge blocking builds.)

**Comment by yujuhong** [comment]
> Do we retry after it times out, or do we have some mechanism to determine if it's already taking too long (for some definition of too long).

We *don't* set a timeout for image pulling as of now. Depending on the image and the network bandwidth, it can take arbitrarily long time to pull an image. There has been discussion to either
 1. provide a `--image-pull-timeout` flag in kubelet to set an absolute deadline, or
 2. periodically check the image pulling progress, and cancel if it's not making progress (or if it's too slow); probably need a configurable flag as well.

It *may* help to implement one of the options, but I don't think any of these can solve the problem in our tests entirely. I've seen slow image pull timeout and being retried, and kept timing out in tests (during a window when we had a 2-min timeout set for image pulls).

**Comment by yujuhong** [comment]
> @bprashanth - I can't agree with you - registry and docker pull flakes have a lot to do with the kubernetes. We should have a mechanism that detects the problems and retries download - especially for system pods (I have a very strong opinion on divergence as well - exactly the same as I had with master registration)

Yes, we can detect problems and retry, but what if the registry is just slow? The test will likely time out before the pull succeed. I don't think we should stress test the image registry in all our e2e tests, even if we want to test it. A local registry seems like a better compromise, although it does make our test cluster further deviate from the prod environment.

**Comment by bprashanth** [comment]
@zmerlynn meaning pipe it as an env in specific test configs? That sounds ok too, but we'd pipe it through all the builds (pre and post commit) that can block merges. The timeout that we're hitting is the per test timeout of 5-10m, when they poll for pods to become running/ready, meaning even gce-slow can hit this. 

While I agree with keeping a consistent build and test env, it's not like we don't mock out database X when testing database Y interanally. There is still a merge blocking suites that does images pulls (node e2e).

**Comment by Random-Liu** [comment]
> 2. periodically check the image pulling progress, and cancel if it's not making progress (or if it's too slow); probably need a configurable flag as well.

This may not help this issue. In this issue, the image pulling did make progresses, just too slow. :)

**Comment by gmarek** [comment]
@yujuhong - I believe that we should think *how we can solve this problem*, not *how we can solve this problem in tests*. We need a convincing user story for that. Maybe a master should double as a registry for system pods?

**Comment by yujuhong** [comment]
@gmarek we will be doing #26300 to help users. I just wanted to point out that this may not help the tests (which is what this issue is for)

**Comment by gmarek** [comment]
I probably don't understand what you're saying. From my point of view tests *are* users. Ones that actually care what they get. In this particular case, tests want to have all system pods running, which translates to having fully working cluster (where by fully I mean "with all the monitoring and stuff that we ship with it running"). I don't think this is very obscure/strange requirement - I think this is how it should be.

Running custom (user) containers in the cluster is something slightly different. It should be working, and we should do whatever we can to mitigate problems with image pull, but it's not as critical as being able to start (with the definition from previous paragraph) a cluster *at all*. I think that we should find some way to assure that as long as someone is running out of the box version of k8s, the cluster will start up successfully.

**Comment by gmarek** [comment]
@fejta - any reason why you assigned me, @zmerlynn and @yujuhong?

**Comment by fejta** [comment]
Just based on thread activity. Please correct the assignees and/or unassign yourself whenever I do a bad job -- if it goes back to zero assiggnees I'll send to Tim/Dawn to route accordingly.

**Comment by dchen1107** [comment]
cc/ @ojarjur

Based on all above discussion and founds, I don't think the issue is at docker, instead at either storage backend (cc/ @ojarjur), or the network. Briefly chatted with @girishkalele, who is writing a simple program testing the node network connectivity. If that program is ready, we can have more signals.

**Comment by yujuhong** [comment]
> I probably don't understand what you're saying. From my point of view tests are users. Ones that actually care what they get. In this particular case, tests want to have all system pods running, which translates to having fully working cluster (where by fully I mean "with all the monitoring and stuff that we ship with it running"). I don't think this is very obscure/strange requirement - I think this is how it should be.

Ok, I didn't realize you were making the distinction between system pods and user pods. I was trying to tackle the more general problem -- slow/hanging image pulling, regardless of what type of pod it is. If the registry/network is just slow (assuming that's the case), kubelet will retry the request until it can succeed, and users have to wait. This long wait doesn't help the tests.

**Comment by yujuhong** [comment]
> Based on all above discussion and founds, I don't think the issue is at docker, instead at either storage backend (cc/ @ojarjur), or the network. Briefly chatted with @girishkalele, who is writing a simple program testing the node network connectivity. If that program is ready, we can have more signals.

I doubt the problem is docker (although not impossible) because it was only pulling one image on that node at the time. 

However, I was wrong about the node being the only one pulling images. I only checked ~1000 nodes (which I thought was the size of the cluster), but according to https://github.com/kubernetes/kubernetes/issues/26626#issuecomment-223103563, logs for half of the cluster were not uploaded. There is no way to tell what other nodes were doing. It is still possible that this was caused by the high load (2000 nodes pulling images) during the cluster startup.

**Comment by bprashanth** [comment]
> Maybe a master should double as a registry for system pods?

I'm personally not signing up for manaing a download server that's going to get hammered from 1000-2000 nodes simultaneously pulling a 400mb image, on an n1-standard-32 that's already pretty close to maxing out.

**Comment by gmarek** [comment]
@bprashanth - it need not to be a centralized system;)

**Comment by bprashanth** [comment]
I believe that's called gcr.io :)

**Comment by dchen1107** [comment]
@girishkalele I am assigning this one to you since you are debugging on the network connectivity issue here. Once you have the small program ready to detect the network connectivity / slowness issue, we need to make it run on every test nodes.

**Comment by girishkalele** [comment]
@dchen1107 

Almost there - we need to discuss how to package the new binary and run it during node init.

Proposed nethealth functionality

The nethealth binary performs a quick HTTP GET download speed check
Key Features:
   - Shell-script friendly - returns a non-zero exit code on timeout or corruption or bandwidth threshold failures
   - Timeout configurable to abort the test early (for super slow links)
   - Can compare actual bandwidth against a command line minimum bandwidth parameter and return a non-zero exit code.
   - Corruption check - can download a checksum file and compute blob checksum and compare. 
   - Configurable object URL for non-GCE environments

**Comment by bprashanth** [comment]
retitled, since most image pulling was addressed by prepulling. At least we no longer get test flake that we need to dig into only to realize it's image pulling.

**Comment by bprashanth** [comment]
@girishkalele so we're just running your test pod alongside the image puller to detect defunct vms?

**Comment by girishkalele** [comment]
@bprashanth cc@dchen1107 

Yes, as Dawn explained to me, the intent is to run the nethealth container in the static test pod and gather data on GCS download performance and integrate it into standard e2e test log output.

**Comment by girishkalele** [comment]
Missing logs tracked by this
https://github.com/kubernetes/test-infra/issues/100

**Comment by dchen1107** [comment]
@girishkalele Based on our offline discussion, once the nethealth image is build, you can simply update cluster/saltbase/salt/e2e-image-puller/e2e-image-puller.manifest to add a new container to pre-pull pod for testing.

**Comment by yujuhong** [comment]
@girishkalele, are you adding a new static pod in kubelet or adding a new container to the existing image pre-pull pod? The image pre-pull pod has restart policy "Never", which doesn't seem to suit your needs.

**Comment by bprashanth** [comment]
I think we just want the netperf timing output dumped to stdout and printed when image pulling fails

**Comment by girishkalele** [comment]
@yujuhong 

I think adding to the existing image pre-pull pod.

This should cause the static pod to run the nethealth container just once upon startup - it will print the bandwidth measurement to stdout and exit.

**Comment by yujuhong** [comment]
Ah.. I see. I thought it was meant to be run longer. Thanks for the explanation.

**Comment by bprashanth** [comment]
Well at this point it might make sense to just break out the shell script in the image puller, and invoke your netperf to dump its stats inbeween image pulls, assuming it's quick enough. I'm afraind that running exactly once might not be enough. Running repeatedly in the bg might help, but then you'd need to synchronize with the image puller on when to exit. 

Leave it to you since this is something we probably shouldn't over engineer at this point of the release cycle.

**Comment by girishkalele** [comment]
Yes, let me build the container and we can invoke it either with a Docker foreground run in the shell script or as part of the pod.

**Comment by girishkalele** [comment]
https://github.com/kubernetes/node-problem-detector/pull/17

**Comment by dchen1107** [comment]
I raised the priority for this pr, since it is important to diagnose the network connectivity issue.

**Comment by girishkalele** [comment]
GCE scalability test flake - **12 GCE nodes have bandwidth measured less than 10 MiB/sec**.
One has only 5 MiB/sec....took 14.2 seconds to download a 64MB blob from GCS.

https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-scalability/8731

```console
STEP: Logs of kube-system/e2e-image-puller-e2e-scalability-minion-group-5kyj:nethealth-check on node e2e-scalability-minion-group-5kyj
Jun 21 15:06:40.824: INFO:  : STARTLOG
2016/06/21 21:53:58 HTTP HEAD reports content length: 67108864 - running GET
2016/06/21 21:54:13 DOWNLOAD: 67108864 bytes 14256 ms Bandwidth ~ 4597 KiB/sec
2016/06/21 21:54:13 ERROR: Minimum bandwidth guarantee of 10 MiB/sec not met - network connectivity is slow
```

```console
STEP: Logs of kube-system/e2e-image-puller-e2e-scalability-minion-group-wi9x:nethealth-check on node e2e-scalability-minion-group-wi9x
Jun 21 15:07:53.425: INFO:  : STARTLOG
2016/06/21 21:54:06 HTTP HEAD reports content length: 67108864 - running GET
2016/06/21 21:54:14 DOWNLOAD: 67108864 bytes 8045 ms Bandwidth ~ 8146 KiB/sec
2016/06/21 21:54:14 ERROR: Minimum bandwidth guarantee of 10 MiB/sec not met - network connectivity is slow
```
and
```
STEP: Logs of kube-system/e2e-image-puller-e2e-scalability-minion-group-ohil:nethealth-check on node e2e-scalability-minion-group-ohil
Jun 21 15:07:45.424: INFO:  : STARTLOG
2016/06/21 21:53:55 HTTP HEAD reports content length: 67108864 - running GET
2016/06/21 21:54:02 DOWNLOAD: 67108864 bytes 7173 ms Bandwidth ~ 9136 KiB/sec
2016/06/21 21:54:02 ERROR: Minimum bandwidth guarantee of 10 MiB/sec not met - network connectivity is slow
```

**Comment by girishkalele** [comment]
Closing - The fixes from #27940 and #29782 seem to have fixed the flaky static pod pulling.
